### PR TITLE
Redefining supported GHC versions, dependency lower bounds, and CI

### PR DESCRIPTION
**Pull request by recursion-ninja**
Some minor codebase modernization:
 - Adding CI workflow for Github Actions. 
 - Removing very old GHC version support
 - Removing CPP usage
 - Cleaning up HLint hints
 - Updating dependency lower bounds

### CODE REVIEW

File: examples/src/Email.hs [hunk 1]
```diff
@@ -4,6 +4,7 @@
 -- Thanks to @purefn for the help on this!
 
 import Control.Lens
+--import Data.Functor ((<$))
```
Comment (gwils) [inline examples/src/Email.hs:7]: Please delete this line instead of commenting it out.

File: lower-bounds.project [hunk 2]
```diff
@@ -0,0 +1,15 @@
+constraints:
```
Comment (gwils) [inline lower-bounds.project:1]: Why is this file necessary? Won't cabal select the correct bounds for a given version on its own?

**Review by gwils — CHANGES_REQUESTED** [review]

File: lower-bounds.project (hunk 2 above)
Comment (recursion-ninja) [reply lower-bounds.project:1]: Yes, cabal will select the correct bounds without specifying this file as the "project file." However, cabal selects the *newest* version of each dependency possible by default. When this file is specified as the "project file" it redefines the constraints to force cabal to choose the *lowest* version of each dependency.

This file is only used in the CI to ensure that `validation` does in fact build when the lowest version of each dependency is selected.

I can remove this from the commit and the"Lower Bounds Check" from the GitHub Actions CI if you would like.

File: examples/src/Email.hs (hunk 1 above)
Comment (recursion-ninja) [reply examples/src/Email.hs]: Good catch. I removed it.

File: lower-bounds.project (hunk 2 above)
Comment (gwils) [reply lower-bounds.project:1]: That's very cool and useful! Thank you for explaining it to me.

**Review by gwils — APPROVED** [review]

**Comment by gwils** [comment]
This looks great! Now unfortunately, it looks like there are merge conflicts you will need to fix.

**Comment by recursion-ninja** [comment]
@gwils I think it's good to go now.

**Comment by gwils** [comment]
@recursion-ninja an error occurred - can you please have a look?
https://github.com/system-f/validation/actions/runs/623493529/workflow

**Comment by recursion-ninja** [comment]
@gwils Wish GitHub has a way to do a "dry run" of the workflow before committing. It would have caught this. See #50 .
